File: ECG BP Monitoring/ha-addon/ecg2_monitor/app/dsp.py
"""Pure-Python DSP for the ecg2 stream: IIR biquads, mains notch, band-pass,
a Pan-Tompkins-style QRS detector and a simple signal-quality estimate.
No numpy/scipy so the image stays small on armv7/aarch64 Alpine."""
from __future__ import annotations

import math
from collections import deque


class Biquad:
    """Direct-form-II transposed biquad."""

    __slots__ = ("b0", "b1", "b2", "a1", "a2", "z1", "z2")

    def __init__(self, b0, b1, b2, a0, a1, a2):
        self.b0, self.b1, self.b2 = b0 / a0, b1 / a0, b2 / a0
        self.a1, self.a2 = a1 / a0, a2 / a0
        self.z1 = self.z2 = 0.0

    def process(self, x: float) -> float:
        y = self.b0 * x + self.z1
        self.z1 = self.b1 * x - self.a1 * y + self.z2
        self.z2 = self.b2 * x - self.a2 * y
        return y

    def reset(self):
        self.z1 = self.z2 = 0.0

    # --- RBJ cookbook designs -------------------------------------------
    @classmethod
    def notch(cls, fs: float, f0: float, q: float = 30.0) -> "Biquad":
        w0 = 2 * math.pi * f0 / fs
        alpha = math.sin(w0) / (2 * q)
        c = math.cos(w0)
        return cls(1, -2 * c, 1, 1 + alpha, -2 * c, 1 - alpha)

    @classmethod
    def lowpass(cls, fs: float, fc: float, q: float = 0.7071) -> "Biquad":
        w0 = 2 * math.pi * fc / fs
        alpha = math.sin(w0) / (2 * q)
        c = math.cos(w0)
        return cls((1 - c) / 2, 1 - c, (1 - c) / 2, 1 + alpha, -2 * c, 1 - alpha)

    @classmethod
    def highpass(cls, fs: float, fc: float, q: float = 0.7071) -> "Biquad":
        w0 = 2 * math.pi * fc / fs
        alpha = math.sin(w0) / (2 * q)
        c = math.cos(w0)
        return cls((1 + c) / 2, -(1 + c), (1 + c) / 2, 1 + alpha, -2 * c, 1 - alpha)


class EcgFilter:
    """Display filter chain: HP 0.5 Hz (baseline wander) -> mains notch (+ 2nd
    harmonic) -> LP 40 Hz. Returns the filtered sample."""

    def __init__(self, fs: float, mains_hz: int = 50):
        self.fs = fs
        self.stages = [
            Biquad.highpass(fs, 0.5),
            Biquad.notch(fs, mains_hz, 25.0),
            Biquad.notch(fs, mains_hz * 2, 25.0),
            Biquad.lowpass(fs, 40.0),
        ]

    def process(self, x: float) -> float:
        for s in self.stages:
            x = s.process(x)
        return x

    def reset(self):
        for s in self.stages:
            s.reset()


class QrsDetector:
    """Pan-Tompkins-lite. Polarity independent (squares the derivative), so a
    lead that shows QRS as negative spikes (as the ecg2 does) still works.

    Feed *filtered* samples via process(); it returns True on a detected beat.
    hr_bpm / rr_ms expose the running estimate."""

    def __init__(self, fs: float):
        self.fs = fs
        self.bp = [Biquad.highpass(fs, 5.0), Biquad.lowpass(fs, 20.0)]
        self.win = max(1, int(0.150 * fs))          # 150 ms integration window
        self.refractory = int(0.300 * fs)           # 300 ms (max 200 bpm)
        self.buf = deque(maxlen=self.win)
        self.integ_sum = 0.0
        self.prev = 0.0
        self.n = 0
        self.last_peak_n = -10 ** 9
        self.spk = 0.0      # running signal peak
        self.npk = 0.0      # running noise peak
        self.thr = 0.0
        self.rr = deque(maxlen=8)
        self.hr_bpm: float | None = None
        self.rr_ms: float | None = None
        self.beats = 0
        self._local_max = 0.0
        self._local_max_n = 0
        self._rising = False

    def reset(self):
        self.__init__(self.fs)

    def process(self, x: float) -> bool:
        self.n += 1
        for s in self.bp:
            x = s.process(x)
        d = x - self.prev
        self.prev = x
        e = d * d
        if len(self.buf) == self.buf.maxlen:
            self.integ_sum -= self.buf[0]
        self.buf.append(e)
        self.integ_sum += e
        y = self.integ_sum / self.win

        beat = False
        # warm-up: learn thresholds for the first 2 s
        if self.n < 2 * self.fs:
            self.spk = max(self.spk * 0.999, y)
            self.thr = 0.25 * self.spk
            return False

        # peak tracking: detect a local maximum of the integrated signal
        if y > self._local_max:
            self._local_max = y
            self._local_max_n = self.n
            self._rising = True
        elif self._rising and y < 0.5 * self._local_max:
            # local max is over — classify it
            pk, pk_n = self._local_max, self._local_max_n
            self._local_max = 0.0
            self._rising = False
            if pk_n - self.last_peak_n > self.refractory and pk > self.thr:
                if self.last_peak_n > 0:
                    rr = (pk_n - self.last_peak_n) / self.fs
                    if 0.3 <= rr <= 2.0:
                        self.rr.append(rr)
                        self.rr_ms = rr * 1000
                        self.hr_bpm = 60.0 / (sum(self.rr) / len(self.rr))
                self.last_peak_n = pk_n
                self.spk = 0.125 * pk + 0.875 * self.spk
                self.beats += 1
                beat = True
            else:
                self.npk = 0.125 * pk + 0.875 * self.npk
            self.thr = self.npk + 0.25 * (self.spk - self.npk)
        # missed-beat search-back: if no beat for 1.66x mean RR, lower threshold
        if self.rr and (self.n - self.last_peak_n) > 1.66 * (sum(self.rr) / len(self.rr)) * self.fs:
            self.thr *= 0.5
        return beat


class SignalQuality:
    """Cheap quality estimate over a sliding 2 s window of *raw* samples:
    flat line / rail saturation / excessive high-frequency noise."""

    def __init__(self, fs: float):
        self.fs = fs
        self.win = deque(maxlen=int(2 * fs))
        self.diff = deque(maxlen=int(2 * fs))
        self.prev = None

    def push(self, x: float):
        self.win.append(x)
        if self.prev is not None:
            self.diff.append(abs(x - self.prev))
        self.prev = x

    def assess(self) -> tuple[str, float]:
        """Returns (label, score 0..1)."""
        if len(self.win) < self.fs:
            return "warming", 0.0
        lo, hi = min(self.win), max(self.win)
        span = hi - lo
        if span < 20:
            return "flat", 0.0
        if lo <= -32000 or hi >= 32000:
            return "saturated", 0.1
        mean_diff = sum(self.diff) / max(1, len(self.diff))
        # ratio of sample-to-sample jitter to overall span: high => noisy
        ratio = mean_diff / span
        if ratio > 0.12:
            return "noisy", max(0.0, 0.6 - ratio)
        if ratio > 0.06:
            return "fair", 0.7
        return "good", 1.0

File: ECG BP Monitoring/android-app/AetosOneHealth/app/src/main/java/com/aetostechlabs/aetosonehealth/dsp/Dsp.kt
package com.aetostechlabs.aetosonehealth.dsp

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.PI

/** Direct-form-II transposed biquad. Ported from the add-on's dsp.py. */
class Biquad(b0: Double, b1: Double, b2: Double, a0: Double, a1: Double, a2: Double) {
    private val b0 = b0 / a0
    private val b1 = b1 / a0
    private val b2 = b2 / a0
    private val a1 = a1 / a0
    private val a2 = a2 / a0
    private var z1 = 0.0
    private var z2 = 0.0

    fun process(x: Double): Double {
        val y = b0 * x + z1
        z1 = b1 * x - a1 * y + z2
        z2 = b2 * x - a2 * y
        return y
    }

    fun reset() { z1 = 0.0; z2 = 0.0 }

    companion object {
        fun notch(fs: Double, f0: Double, q: Double = 30.0): Biquad {
            val w0 = 2 * PI * f0 / fs
            val alpha = sin(w0) / (2 * q)
            val c = cos(w0)
            return Biquad(1.0, -2 * c, 1.0, 1 + alpha, -2 * c, 1 - alpha)
        }
        fun lowpass(fs: Double, fc: Double, q: Double = 0.7071): Biquad {
            val w0 = 2 * PI * fc / fs
            val alpha = sin(w0) / (2 * q)
            val c = cos(w0)
            return Biquad((1 - c) / 2, 1 - c, (1 - c) / 2, 1 + alpha, -2 * c, 1 - alpha)
        }
        fun highpass(fs: Double, fc: Double, q: Double = 0.7071): Biquad {
            val w0 = 2 * PI * fc / fs
            val alpha = sin(w0) / (2 * q)
            val c = cos(w0)
            return Biquad((1 + c) / 2, -(1 + c), (1 + c) / 2, 1 + alpha, -2 * c, 1 - alpha)
        }
    }
}

/** HP 0.5 Hz (baseline wander) -> mains notch (+2nd harmonic) -> LP 40 Hz. */
class EcgFilter(fs: Double, mainsHz: Int = 50) {
    private val stages = listOf(
        Biquad.highpass(fs, 0.5),
        Biquad.notch(fs, mainsHz.toDouble(), 25.0),
        Biquad.notch(fs, mainsHz * 2.0, 25.0),
        Biquad.lowpass(fs, 40.0)
    )
    fun process(x: Double): Double {
        var v = x
        for (s in stages) v = s.process(v)
        return v
    }
    fun reset() = stages.forEach { it.reset() }
}

/**
 * Pan-Tompkins-lite. Polarity independent (it squares the derivative), so the
 * ecg2's negative-going QRS spikes are detected as well as positive ones.
 */
class QrsDetector(private val fs: Double) {
    private var bp = listOf(Biquad.highpass(fs, 5.0), Biquad.lowpass(fs, 20.0))
    private val win = maxOf(1, (0.150 * fs).toInt())
    private val refractory = (0.300 * fs).toInt()
    private val buf = ArrayDeque<Double>(win)
    private var integSum = 0.0
    private var prev = 0.0
    private var n = 0
    private var lastPeakN = -1_000_000_000
    private var spk = 0.0
    private var npk = 0.0
    private var thr = 0.0
    private val rr = ArrayDeque<Double>(8)
    private var localMax = 0.0
    private var localMaxN = 0
    private var rising = false

    var hrBpm: Double? = null; private set
    var rrMs: Double? = null; private set
    var beats = 0; private set

    fun reset() {
        bp = listOf(Biquad.highpass(fs, 5.0), Biquad.lowpass(fs, 20.0))
        buf.clear(); rr.clear()
        integSum = 0.0; prev = 0.0; n = 0; lastPeakN = -1_000_000_000
        spk = 0.0; npk = 0.0; thr = 0.0
        localMax = 0.0; localMaxN = 0; rising = false
        hrBpm = null; rrMs = null; beats = 0
    }

    fun process(sample: Double): Boolean {
        n++
        var x = sample
        for (s in bp) x = s.process(x)
        val d = x - prev
        prev = x
        val e = d * d
        if (buf.size == win) integSum -= buf.removeFirst()
        buf.addLast(e)
        integSum += e
        val y = integSum / win

        // warm-up: learn thresholds over the first 2 s
        if (n < 2 * fs) {
            spk = maxOf(spk * 0.999, y)
            thr = 0.25 * spk
            return false
        }

        var beat = false
        if (y > localMax) {
            localMax = y; localMaxN = n; rising = true
        } else if (rising && y < 0.5 * localMax) {
            val pk = localMax
            val pkN = localMaxN
            localMax = 0.0; rising = false
            if (pkN - lastPeakN > refractory && pk > thr) {
                if (lastPeakN > 0) {
                    val interval = (pkN - lastPeakN) / fs
                    if (interval in 0.3..2.0) {
                        if (rr.size == 8) rr.removeFirst()
                        rr.addLast(interval)
                        rrMs = interval * 1000
                        hrBpm = 60.0 / (rr.sum() / rr.size)
                    }
                }
                lastPeakN = pkN
                spk = 0.125 * pk + 0.875 * spk
                beats++
                beat = true
            } else {
                npk = 0.125 * pk + 0.875 * npk
            }
            thr = npk + 0.25 * (spk - npk)
        }
        // missed-beat search-back
        if (rr.isNotEmpty() && (n - lastPeakN) > 1.66 * (rr.sum() / rr.size) * fs) thr *= 0.5
        return beat
    }
}

/** Flat line / rail saturation / high-frequency noise over a 2 s window. */
class SignalQuality(private val fs: Double) {
    private val cap = (2 * fs).toInt()
    private val win = ArrayDeque<Double>(cap)
    private val diff = ArrayDeque<Double>(cap)
    private var prev: Double? = null

    fun push(x: Double) {
        if (win.size == cap) win.removeFirst()
        win.addLast(x)
        prev?.let {
            if (diff.size == cap) diff.removeFirst()
            diff.addLast(abs(x - it))
        }
        prev = x
    }

    fun assess(): Pair<String, Double> {
        if (win.size < fs) return "warming" to 0.0
        val lo = win.min(); val hi = win.max()
        val span = hi - lo
        if (span < 20) return "flat" to 0.0
        if (lo <= -32000 || hi >= 32000) return "saturated" to 0.1
        val meanDiff = diff.sum() / maxOf(1, diff.size)
        val ratio = meanDiff / span
        return when {
            ratio > 0.12 -> "noisy" to maxOf(0.0, 0.6 - ratio)
            ratio > 0.06 -> "fair" to 0.7
            else -> "good" to 1.0
        }
    }
}
